src/components/abstractComponent.ts

```diff
@@ -0,0 +1,579 @@
+
+  protected _content: D;
+  protected _destroyed = false;
+  protected _height: number;
+  protected _isAnchored = false;
+  protected _isSetup = false;
+  protected _onAnchorCallbacks: Utils.CallbackSet<GenericComponentCallback<D>>
+    = new Utils.CallbackSet<GenericComponentCallback<D>>();
+  protected _onDetachCallbacks: Utils.CallbackSet<GenericComponentCallback<D>>
+    = new Utils.CallbackSet<GenericComponentCallback<D>>();
+  protected _origin: Point = { x: 0, y: 0 };
+  protected _resizeHandler: IResizeHandler;
+  protected _width: number;
+  protected _xAlignment: string = "left";
+  protected _yAlignment: string = "top";
+
+  protected static _xAlignToProportion: { [alignment: string]: number } = {
+    "left": 0,
+    "center": 0.5,
+    "right": 1,
+  };
+  protected static _yAlignToProportion: { [alignment: string]: number } = {
+    "top": 0,
+    "center": 0.5,
+    "bottom": 1,
+  };
+
+  /**
+   * Attaches the Component as a child of a given html element.
+   *
+   * @param {D} selection.
+   * @returns {Component} The calling Component.
+   */
+  abstract anchor(selection: D): this;
+  abstract anchorHTML(selection: HTMLElement): this;
+  /**
+   * Adds a callback to be called on anchoring the Component to the DOM.
+   * If the Component is already anchored, the callback is called immediately.
+   *
+   * @param {GenericComponentCallback} callback
+   * @return {Component}
+   */
+  public onAnchor(callback: GenericComponentCallback<D>) {
+    if (this._isAnchored) {
+      callback(this);
+    }
+    this._onAnchorCallbacks.add(callback);
+    return this;
+  }
+  /**
+   * Removes a callback that would be called on anchoring the Component to the DOM.
+   * The callback is identified by reference equality.
+   *
+   * @param {GenericComponentCallback} callback
+   * @return {Component}
+   */
+  public offAnchor(callback: GenericComponentCallback<D>) {
+    this._onAnchorCallbacks.delete(callback);
+    return this;
+  }
+  /**
+   * Given available space in pixels, returns the minimum width and height this Component will need.
+   *
+   * @param {number} availableWidth
+   * @param {number} availableHeight
+   * @returns {SpaceRequest}
+   */
+  public requestedSpace(availableWidth: number, availableHeight: number) {
+    return {
+      minWidth: 0,
+      minHeight: 0,
+    };
+  }
+  /**
+   * Computes and sets the size, position, and alignment of the Component from the specified values.
+   * If no parameters are supplied and the Component is a root node,
+   * they are inferred from the size of the Component's element.
+   *
+   * @param {Point} [origin] Origin of the space offered to the Component.
+   * @param {number} [availableWidth] Available width in pixels.
+   * @param {number} [availableHeight] Available height in pixels.
+   * @returns {IComponent} The calling Component.
+   */
+  abstract computeLayout(origin?: Point, availableWidth?: number, availableHeight?: number): this;
```

Comment:
The `computeLayout` method is already defined through `IComponent` interface so it should automatically turn into an abstract method on this class right?